core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -4474,6 +4474,21 @@ protected void validateWhereOrOn(
       SqlNode condition,
       String clause) {
     validateNoAggs(aggOrOverOrGroupFinder, condition, clause);
+    //  SqlSelect need to expand alias
+    condition.accept(new SqlShuttle() {
+      @Override public @Nullable SqlNode visit(SqlCall call) {
+        call.getOperandList()
+            .stream()
+            .filter(Objects::nonNull)
+            .forEach(node -> node.accept(this));
+        if (call.getKind() == SqlKind.SELECT) {
+          SqlSelect select = (SqlSelect) call;
+          validateHavingClause(select);
```

Comment:
`validateWhereOrOn` is used not only to validate `where` clause, is your change also applies to `on` clause, is there any test covers this?

Reply:
Thanks for your review, I added one test.